File: 2024/rust/src/main.rs
mod aoc;

fn main() {
    aoc::run("../input/day1", day1)
}

fn day1(input: &str) {
    let mut left_list: [i32; 1000] = [0; 1000];
    let mut right_list: [i32; 1000] = [0; 1000];
    let mut i = 0;
    for line in input.lines() {
        let left_num: i32 = line[0..5].parse().unwrap();
        let right_num: i32 = line[8..].parse().unwrap();
        left_list[i] = left_num;
        right_list[i] = right_num;
        i += 1;
    }
    left_list.sort_unstable();
    right_list.sort_unstable();
    let mut part1_res = 0;
    let mut cnt_map: [u8; 99999] = [0; 99999];
    for i in 0..1000 {
        let a = left_list[i];
        let b = right_list[i];
        cnt_map[b as usize] += 1;
        // println!("{:?}", entry.);
        part1_res += (a - b).abs();
    }
    let mut part2_res = 0;
    for num in left_list {
        part2_res += num * cnt_map[num as usize] as i32;
    }
    println!("part1: {}, part2: {}", part1_res, part2_res);
}

File: 2024/lua/main.lua
dofile("aoc.lua")

function SplitLines(input)
    local lines = {}
    for line in string.gmatch(input, "[^\n]+") do
        table.insert(lines, line)
    end
    return lines
end

function PrintArray(t)
    io.write("[")
    for key, value in pairs(t) do
        if type(value) == "table" then
            io.write(key .. ":")
            PrintArray(value) -- Recursively print nested tables
        else
            io.write(value .. ", ")
        end
    end
    io.write("]\n")
end

function Day1(input)
    local left_list = {}
    local right_list = {}
    local lines = SplitLines(input)
    for _, line in ipairs(lines) do
        if line == "" then break end;
        local left_num = tonumber(string.sub(line, 0, 5));
        local right_num = tonumber(string.sub(line, 8));
        table.insert(left_list, left_num)
        table.insert(right_list, right_num)
    end
    table.sort(left_list)
    table.sort(right_list)

    local part1_res = 0;
    local cnt_map = {}
    for i = 1, #left_list do
        PrintArray(left_list)
        local a = left_list[i]
        local b = right_list[i]
        cnt_map[b] = (cnt_map[b] or 0) + 1
        part1_res = part1_res + math.abs(a - b)
    end

    local part2_res = 0
    for _, num in ipairs(left_list) do
        part2_res = part2_res + (cnt_map[num] or 0) * num;
    end
    print("part1: " .. part1_res .. " part2: " .. part2_res);
end

Run("../input/day1", Day1)
